clang/runtime/dpct-rt/include/dpct/sparse_utils.hpp

```diff
@@ -1458,6 +1520,243 @@ inline void csr2csc(sycl::queue queue, int m, int n, int nnz,
+  const auto Iter = desc->get_csrgemm_info_map().find(args);
+
+  oneapi::mkl::sparse::matrix_handle_t *matrix_handle_a;
+  oneapi::mkl::sparse::matrix_handle_t *matrix_handle_b;
+  oneapi::mkl::sparse::matrix_handle_t *matrix_handle_c;
+  oneapi::mkl::sparse::matmat_descr_t *matmat_desc;
+
+#ifdef DPCT_USM_LEVEL_NONE
+  auto row_ptr_a_buf = dpct::detail::get_memory<int>(row_ptr_a);
+  auto col_ind_a_buf = dpct::detail::get_memory<int>(col_ind_a);
+  auto val_a_buf = dpct::detail::get_memory<Ty>(val_a);
+  auto row_ptr_b_buf = dpct::detail::get_memory<int>(row_ptr_b);
+  auto col_ind_b_buf = dpct::detail::get_memory<int>(col_ind_b);
+  auto val_b_buf = dpct::detail::get_memory<Ty>(val_b);
+  auto row_ptr_c_buf = dpct::detail::get_memory<int>(row_ptr_c);
+  auto col_ind_c_buf = dpct::detail::get_memory<int>(nullptr);
+  auto val_c_buf = dpct::detail::get_memory<Ty>(nullptr);
+#endif
+
+  if (Iter == desc->get_csrgemm_info_map().end()) {
+    matrix_handle_a = new oneapi::mkl::sparse::matrix_handle_t();
+    matrix_handle_b = new oneapi::mkl::sparse::matrix_handle_t();
+    matrix_handle_c = new oneapi::mkl::sparse::matrix_handle_t();
+    oneapi::mkl::sparse::init_matrix_handle(matrix_handle_a);
+    oneapi::mkl::sparse::init_matrix_handle(matrix_handle_b);
+    oneapi::mkl::sparse::init_matrix_handle(matrix_handle_c);
+    int rows_c = m;
+    int cols_c = n;
+    int rows_a = (trans_a == oneapi::mkl::transpose::nontrans) ? m : k;
+    int cols_a = (trans_a == oneapi::mkl::transpose::nontrans) ? k : m;
+    int rows_b = (trans_b == oneapi::mkl::transpose::nontrans) ? k : n;
+    int cols_b = (trans_b == oneapi::mkl::transpose::nontrans) ? n : k;
+
+#ifdef DPCT_USM_LEVEL_NONE
+    oneapi::mkl::sparse::set_csr_data(queue, *matrix_handle_a, rows_a, cols_a,
+                                      info_a->get_index_base(), row_ptr_a_buf,
+                                      col_ind_a_buf, val_a_buf);
+    oneapi::mkl::sparse::set_csr_data(queue, *matrix_handle_b, rows_b, cols_b,
+                                      info_b->get_index_base(), row_ptr_b_buf,
+                                      col_ind_b_buf, val_b_buf);
+    oneapi::mkl::sparse::set_csr_data(queue, *matrix_handle_c, rows_c, cols_c,
+                                      info_c->get_index_base(), row_ptr_c_buf,
+                                      col_ind_c_buf, val_c_buf);
+#else
+    oneapi::mkl::sparse::set_csr_data(
+        queue, *matrix_handle_a, rows_a, cols_a, info_a->get_index_base(),
+        (int *)row_ptr_a, (int *)col_ind_a, (Ty *)val_a);
+    oneapi::mkl::sparse::set_csr_data(
+        queue, *matrix_handle_b, rows_b, cols_b, info_b->get_index_base(),
+        (int *)row_ptr_b, (int *)col_ind_b, (Ty *)val_b);
+    oneapi::mkl::sparse::set_csr_data(queue, *matrix_handle_c, rows_c, cols_c,
+                                      info_c->get_index_base(), row_ptr_c,
+                                      (int *)nullptr, (Ty *)nullptr);
```

Comment:
You may want to keep a remark here for future that oneMKL would need to eventually allow `(int *)nullptr` being passed in even for `row_ptr_c` in the initial calls before `matmat`. This would make it consistent with cuSPARSE, but right now it isn't.

Reply:
OK. I will add a comment.